src/omv/ports/stm32/omv_gpu.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * This file is part of the OpenMV project.
+ *
+ * Copyright (c) 2013-2024 Ibrahim Abdelkader <iabdalkader@openmv.io>
+ * Copyright (c) 2013-2024 Kwabena W. Agyeman <kwagyeman@openmv.io>
+ *
+ * This work is licensed under the MIT license, see the file LICENSE for details.
+ *
+ * GPU driver for STM32 port.
+ */
+#include "omv_boardconfig.h"
+#if (OMV_GPU_ENABLE == 1)
+#include "imlib.h"
+
+#include STM32_HAL_H
+#include "dma.h"
+
+#define TIME_GPU    (0)
+#if (TIME_GPU == 1)
+#include "py/mphal.h"
+#include <stdio.h>
+#endif
+
+bool omv_gpu_draw_image(image_t *dst_img,
+                        rectangle_t *dst_rect,
+                        image_t *src_img,
+                        rectangle_t *src_rect,
+                        int alpha,
+                        const uint16_t *color_palette,
+                        const uint8_t *alpha_palette,
+                        image_hint_t hint) {
+    // DMA2D can only draw on RGB565 buffers and the destination/source buffers must be accessible by DMA.
+    if ((dst_img->pixfmt != PIXFORMAT_RGB565) || (!DMA_BUFFER(dst_img->data)) || (!DMA_BUFFER(src_img->data))) {
+        return false;
+    }
+
+    // DMA2D can only read from RGB565 or GRAYSCALE buffers.
+    // If the source image is RGB565, it must not have a color or alpha palette.
+    if ((src_img->pixfmt != PIXFORMAT_GRAYSCALE) &&
+        ((src_img->pixfmt != PIXFORMAT_RGB565) || color_palette || alpha_palette)) {
+        return false;
+    }
+
+    // DMA2D cannot scale so ensure that the source and destination rectangles are the same size.
+    // The bilinear/nearest-neighbor flag doesn't matter since we can't scale.
+    if ((dst_rect->w != src_rect->w) || (dst_rect->h != src_rect->h)) {
+        return false;
+    }
+
+    // DMA2D cannot hmirror or vflip.
+    if (hint & (IMAGE_HINT_HMIRROR | IMAGE_HINT_VFLIP)) {
+        return false;
+    }
+
+    // DMA2D must always fetch the background on the F4 and F7 series so do this in software.
+    #if defined(MCU_SERIES_F4) || defined(MCU_SERIES_F7)
+    if (hint & IMAGE_HINT_BLACK_BACKGROUND) {
+        return false;
+    }
+    #endif
+
+    #if (TIME_GPU == 1)
+    mp_uint_t start = mp_hal_ticks_ms();
+    #endif
+
+    DMA2D_HandleTypeDef dma2d = {};
```

Comment:
Unless you really want to zero-initialize the struct, please avoid doing that. The compiler will zero-initialize every member and by most likely calling `memset`. This does not help performance.

Reply:
The goal was to zero the struct here as the config is on the stack. I can move it to a static var.

Reply:
If you set all member then you don't need to zero the struct. Stack is better, as it saves BSS.

Reply:
Done

Reply:
Not all members are set. It's got to be zeroed. So, moving to bss.

Reply:
Please leave to save the BSS, it doesn't need to be global.

Reply:
Put it back on the stack and zeroed it. However, I removed code that was assigning zeroed members constant values of 0. I feel more comfortable zeroing the whole thing since ST adds new fields each CPU.